### PR TITLE
New function : pgr planarFaces

### LINKED ISSUE #3143: New function pgr_planarFaces to be added in pgRouting
## pgr_planarFaces():

**planarFaces()**: Planar face extraction is an algorithm that takes an undirected planar graph, computes a planar embedding of it, and walks every face of that embedding. A face is a maximal connected region of the plane bounded by edges, which on a street network corresponds to the city blocks between the roads, and the exterior face is the unbounded region outside the outermost cycle. Once an embedding is fixed each edge separates exactly two faces, bordering one on its left side and the other on its right, so the function reports every edge-face incidence. This implementation uses the Boost Graph Library's `boyer_myrvold_planarity_test` to compute the embedding and `planar_face_traversal` to walk the faces, with a time complexity of O(V + E), where V is the number of vertices and E is the number of edges. This will enhance pgRouting's capabilities in topology extraction and polygon reconstruction problems.

#### The algorithm:

- Works on **undirected** graphs.
- Edge direction and traversal costs are ignored, only the edge endpoints matter.
- Requires the input graph to be planar, otherwise raises `ERROR: Graph is not planar`. Use `pgr_isPlanar` first when planarity is unknown.
- Reports one row per edge side, so the result has `2|E|` rows.
- Walks the outer face of every connected component separately, so on a graph with `C` components the face count satisfies `|V| - |E| + |F| = 2C`.
- Running time: O(V + E) where V is the number of vertices and E is the number of edges.

### Signature:

- pgr_planarFaces()

```sql
pgr_planarFaces(Edges SQL)

Returns set of (seq, face_id, edge_id, side)
OR EMPTY SET
```

## Parameters

| Parameter | Type | Description |
| --- | --- | --- |
| **Edges SQL** | `TEXT` | Inner SQL query, as described below. |

### Inner Query

**Edges SQL**: An SQL query returning a set of rows with the following columns:

| Column | Type | Default | Description |
| --- | --- | --- | --- |
| **id** | `ANY-INTEGER` | | Identifier of the edge. |
| **source** | `ANY-INTEGER` | | Identifier of the first endpoint vertex of the edge. |
| **target** | `ANY-INTEGER` | | Identifier of the second endpoint vertex of the edge. |
| **cost** | `ANY-NUMERICAL` | | Weight of the edge `(source, target)`. When negative, the edge does not exist. |
| **reverse_cost** | `ANY-NUMERICAL` | `-1` | Weight of the edge `(target, source)`. When negative, the edge does not exist. |

Where:

- `ANY-INTEGER` = `SMALLINT`, `INTEGER`, `BIGINT`
- `ANY-NUMERICAL` = `SMALLINT`, `INTEGER`, `BIGINT`, `REAL`, `FLOAT`

### Result Columns

Returns `SETOF (seq, face_id, edge_id, side)`.

| Column | Type | Description |
| --- | --- | --- |
| **seq** | `BIGINT` | Sequential value starting from 1. |
| **face_id** | `BIGINT` | Identifier of the face in the computed embedding. |
| **edge_id** | `BIGINT` | Identifier of the edge that borders the face. |
| **side** | `INTEGER` | `1` when the edge borders the face on the left side, `2` when on the right side. |

### PR DESCRIPTION
**Pull request by sakirr05**
This PR adds one new function named pgr_planarFaces to pgrouting:
- ``pgr_planarFaces``: Returns the faces of a planar embedding of an undirected graph

This is a final merge PR containing commits from several different PRs combined.

The intermediate PRs made on the [GSoC-pgRouting](https://github.com/pgRouting/GSoC-pgRouting/) repository are as follows:

- pgRouting/GSoC-pgRouting#551
- pgRouting/GSoC-pgRouting#556
- pgRouting/GSoC-pgRouting#559
- pgRouting/GSoC-pgRouting#563
- pgRouting/GSoC-pgRouting#568
- pgRouting/GSoC-pgRouting#577
- pgRouting/GSoC-pgRouting#578
- pgRouting/GSoC-pgRouting#591
- pgRouting/GSoC-pgRouting#595
- pgRouting/GSoC-pgRouting#600
- pgRouting/GSoC-pgRouting#603
- pgRouting/GSoC-pgRouting#611
- pgRouting/GSoC-pgRouting#614
- pgRouting/GSoC-pgRouting#621
- pgRouting/GSoC-pgRouting#627

@pgRouting/admins

## Summary by CodeRabbit

* **New Features**
  * Added the experimental `pgr_planarFaces` function for identifying planar graph faces and their associated edges and sides.
  * Added support for empty inputs, subgraphs, face grouping, and validation of non-planar graphs.
* **Documentation**
  * Added comprehensive planar-family documentation, examples, diagrams, and release-note coverage.
* **Tests**
  * Added query and regression tests covering edge cases, output types, Euler formulas, error handling, and crash prevention.

### CODE REVIEW

**Review by cvvergara — APPROVED** [review]
Thanks for your work.
This approval is about work review.
To be merged it needs approval from PSC with a motion on Discourse.
Any of the other members listed for review, once approved will do.

File: include/drivers/planar_driver.hpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/*PGR-GNU*****************************************************************
+File: planar_driver.hpp
+
+Copyright (c) 2026-2026 pgRouting developers
+Mail: project@pgrouting.org
+
+Design of one process & driver file by
+Copyright (c) 2025 Celia Virginia Vergara Castillo
+Mail: vicky at erosion.dev
```
Comment (krashish8) [inline include/drivers/planar_driver.hpp:9]: I think you should put your own name/email here instead of Vicky's. Also, it's 2026 (and almost the end of the year) 😄

File: include/process/planar_process.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*PGR-GNU*****************************************************************
+File: planar_process.h
+
+Copyright (c) 2026-2026 pgRouting developers
+Mail: project@pgrouting.org
+
+Design of one process & driver file by
+Copyright (c) 2025 Celia Virginia Vergara Castillo
+Mail: vicky at erosion.dev
```
Comment (krashish8) [inline include/process/planar_process.h:9]: same here and below files

**Review by krashish8 — APPROVED** [review]
Looks good!
Thanks for all the efforts 🎉

**Review by sanak — APPROVED** [review]

File: include/process/planar_process.h (hunk 2 above)
Comment (sakirr05) [reply include/process/planar_process.h]: done:)

File: include/drivers/planar_driver.hpp (hunk 1 above)
Comment (sakirr05) [reply include/drivers/planar_driver.hpp]: done:)

**Comment by iosefa** [comment]
I approve adding this function; thank you! But I ran a test in `psql` and found a problem with unequal directional costs. A simple triangle with `cost = 1, reverse_cost = 1` gives 2 faces and 6 rows. Changing `reverse_cost` to 2 gives 5 faces and 12 rows.

The tests also use `sign()` on both costs, which makes all positive values equal to 1 and misses this case. Since costs are supposed to be ignored, I’d expect the same result. It doesn’t crash, but this looks related to the issue I flagged in #3134. I think it’s worth fixing before merging.

**Comment by cvvergara** [comment]
@iosefa can you post the small test you ran please
